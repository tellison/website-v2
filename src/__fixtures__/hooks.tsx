import {
  Binary,
  ContributorApiResponse,
  Contributor,
  EventAPI,
  LatestTemurin,
  MarketplaceRelease,
  MockTemurinFeatureReleaseAPI,
  MockTemurinReleaseAPI,
  News,
  NewsResponse,
  ReleaseAsset,
  TemurinReleases,
} from '../hooks';

export const createRandomTemurinRelease = (installer): ReleaseAsset => ({
  os: 'os_mock',
  architecture: 'arch_mock',
  release_link: new URL('https://release_link_mock'),
  platform_name: 'platform_name_mock',
  release_name: 'release_name_mock',
  release_date: new Date(Date.UTC(2020, 0, 1)),
  binaries: [
    {
      type: 'type_mock',
      link: new URL('https://link_mock'),
      checksum: 'checksum_mock',
      size: 0,
      extension: 'extension_mock',
      installer_link: installer ? new URL('https://installer_link_mock') : undefined,
      installer_checksum: installer ? 'installer_checksum_mock' : undefined,
      installer_size: installer ? 0 : undefined,
      installer_extension: installer ? 'installer_extension_mock' : undefined,
    },
  ],
})

export const createRandomMarketplaceRelease = (installer, id): MarketplaceRelease => ({
  release_name: 'release_name_mock',
  vendor: 'vendor_mock',
  binary: {
    os: 'os_mock',
    architecture: 'arch_mock',
    distribution: 'distribution_mock',
    image_type: 'type_mock',
    jvm_impl: 'jvm_impl_mock',
    timestamp: new Date(Date.UTC(2020, 0, 1)),
    scm_ref: 'scm_ref_mock',
    openjdk_scm_ref: 'openjdk_scm_ref_mock',
    aqavit_results_link: new URL('https://aqavit_results_link_mock'),
    tck_affidavit_link: new URL('https://tck_affidavit_link_mock'),
    package: {
      name: 'name_mock.tar.gz',
      link: new URL('https://link_mock'),
      sha265sum: `sha265sum_mock${id}`,
      sha256sum_link: new URL('https://sha256sum_link_mock'),
      signature_link: new URL('https://signature_link_mock'),
    },
    installer: installer
      ? [
          {
            name: 'installer_name_mock.msi',
            link: new URL('https://installer_link_mock'),
            sha265sum: 'installer_sha265sum_mock',
            sha256sum_link: new URL('https://installer_sha256sum_link_mock'),
            signature_link: new URL('https://installer_signature_link_mock'),
          },
        ]
      : undefined,
  },
  version: {
    major: 0,
    minor: 0,
    security: 0,
    patch: 0,
    build: 0,
    openjdk_version: 'openjdk_version_mock',
  }
})

export const mockLatestTemurin = (installer): LatestTemurin => ({
  download_count: 0,
  id: 'id_mock',
  release_link: new URL('https://release_link_mock'),
  release_name: 'release_name_mock',
  release_type: 'release_type_mock',
  timestamp: new Date(Date.UTC(2020, 0, 1)),
  updated_at: new Date(Date.UTC(2020, 0, 1)),
  vendor: 'vendor_mock',
  version_data: {
    major: 0,
    minor: 0,
    security: 0,
    patch: 0,
    build: 0,
    openjdk_version: 'openjdk_version_mock',
  },
  binaries: [
    {
      os: 'os_mock',
      architecture: 'arch_mock',
      image_type: 'type_mock',
      heap_size: 'heap_size_mock',
      download_count: 0,
      jvm_impl: 'jvm_impl_mock',
      package: {
        name: 'name_mock.tar.gz',
        link: new URL('https://link_mock'),
        checksum: 'sha265sum_mock',
        checksum_link: new URL('https://sha256sum_link_mock'),
        metadata_link: new URL('https://metadata_link_mock'),
        signature_link: new URL('https://signature_link_mock'),
        size: 0,
        download_count: 0,
      },
      installer: installer
      ?
        {
          name: 'installer_name_mock.msi',
          link: new URL('https://installer_link_mock'),
          checksum: 'installer_sha265sum_mock',
          checksum_link: new URL('https://installer_sha256sum_link_mock'),
          metadata_link: new URL('https://installer_metadata_link_mock'),
          signature_link: new URL('https://installer_signature_link_mock'),
          size: 0,
          download_count: 0,
        }
      : undefined,
    }
  ]
})

export const mockNewsAPI = (): NewsResponse => ({
  news: [
    {
      title: 'news_title_mock',
      body: 'body_mock',
      id: 'id_mock',
      link: new URL('https://link_mock'),
      date: new Date(Date.UTC(2020, 0, 1)),
    }
  ],
  pagination: {
    page: 1,
    pagesize: 5,
    result_start: 1,
    result_end: 5,
    result_size: 5,
    total_result_size: 16
  }
});

export const mockEventsAPI = (): EventAPI => ({
  events: [
    {
      title: 'events_title_mock',
      id: 'id_mock',
      infoLink: new URL('https://link_mock'),
      date: new Date(Date.UTC(2020, 0, 1)),
    }
  ]
});

export const createMockTemurinReleaseAPI = (installer): MockTemurinReleaseAPI => ({
  release_link: new URL('https://release_link_mock'),
  release_name: 'release_name_mock',
  vendor: 'vendor_mock',
  version: {
    major: 0,
    minor: 0,
    security: 0,
    patch: 0,
    build: 0,
    openjdk_version: 'openjdk_version_mock',
  },
  binary: {
    architecture: 'arch_mock',
    project: 'project_mock',
    scm_ref: 'scm_ref_mock',
    download_count: 0,
    updated_at: new Date(Date.UTC(2020, 0, 1)),
    os: 'os_mock',
    image_type: 'jdk',
    jvm_impl: 'jvm_impl_mock',
    heap_size: 'heap_size_mock',
    package: {
      name: 'name_mock.tar.gz',
      link: new URL('https://link_mock'),
      checksum: 'checksum_mock',
      checksum_link: new URL('https://checksum_link_mock'),
      metadata_link: new URL('https://metadata_link_mock'),
      signature_link: new URL('https://signature_link_mock'),
      size: 0,
      download_count: 0,
    },
    installer: installer
      ? {
          name: 'installer_name_mock.msi',
          link: new URL('https://installer_link_mock'),
          checksum: 'installer_checksum_mock',
          checksum_link: new URL('https://installer_checksum_link_mock'),
          metadata_link: new URL('https://installer_metadata_link_mock'),
          signature_link: new URL('https://installer_signature_link_mock'),
          size: 0,
          download_count: 0,
        }
      : undefined,
  }
});

export const createMockTemurinFeatureReleaseAPI = (installer): MockTemurinFeatureReleaseAPI => ({
  id: 'id_mock',
  download_count: 0,
  release_name: 'release_name_mock',
  release_link: new URL('https://release_link_mock'),
  release_type: 'release_type_mock',
  timestamp: new Date(Date.UTC(2020, 0, 1)),
  updated_at: new Date(Date.UTC(2020, 0, 1)),
  vendor: 'vendor_mock',
  version_data: {
    major: 0,
    minor: 0,
    security: 0,
    patch: 0,
    build: 0,
    openjdk_version: 'openjdk_version_mock',
  },
  binaries: [
    {
      os: 'os_mock',
      architecture: 'arch_mock',
      image_type: 'jdk',
      package: {
        name: 'name_mock.tar.gz',
        checksum: 'checksum_mock',
        link: new URL('https://link_mock'),
        size: 0,
      },
      installer: installer
        ? {
          name: 'installer_name_mock.msi',
          checksum: 'installer_checksum_mock',
          link: new URL('https://installer_link_mock'),
          size: 0,
        }
      : undefined,
    }
  ]
});

export const createRandomTemurinReleases = (installer): TemurinReleases => ({
  release_name: 'release_name_mock',
  release_link: new URL('https://release_link_mock'),
  source_url: new URL('https://source_url_mock'),
  timestamp: new Date(Date.UTC(2020, 0, 1)),
  platforms: {
    'platform_mock': {
      assets: [
        {
          os: 'os_mock',
          architecture: 'architecture_mock',
          type: 'type_mock',
          link: new URL('https://link_mock'),
          checksum: 'checksum_mock',
          size: 0,
          extension: 'extension_mock',
          installer_link: installer ? new URL('https://installer_link_mock') : undefined,
          installer_checksum: installer ? 'installer_checksum_mock' : undefined,
          installer_size: installer ? 0 : undefined,
          installer_extension: installer ? 'installer_extension_mock' : undefined,
        },
      ],
    },
  },
})

export const createRandomContributorViewData = (): Contributor => ({
  avatarUri: 'avatarUri_mock',
  profileUri: 'profileUri_mock',
  login: 'login_mock',
  contributionsCount: 0,
  commitsListUri: 'commitsListUri_mock',
  repo: 'reponame_mock',
});

export const createRandomLatestForOSData = (): Binary => ({
  release_name: 'release_name_mock',
  link: new URL('https://link_mock'),
});

export const createRandomNewsAndEventsData = (): News => ({
  news: {
    news: [
      {
        id: 'id_mock',
        title: 'title_mock',
        body: 'body_mock',
        date: new Date(Date.UTC(2020, 0, 1)),
        link: new URL('https://link_mock'),
      },
    ],
    pagination: {
      page: 1,
      pagesize: 5,
      result_start: 1,
      result_end: 5,
      result_size: 5,
      total_result_size: 16,
    }
  },
  events: [
    {
      id: 'id_mock',
      title: 'title_mock',
      infoLink: new URL('https://infoLink_mock'),
      date: new Date(Date.UTC(2020, 0, 1)),
    },
  ]
});

export const createRandomContributorViewData1Contribution = (): Contributor => ({
  avatarUri: 'avatarUri_mock',
  profileUri: 'profileUri_mock',
  login: 'login_mock',
  contributionsCount: 1,
  commitsListUri: 'commitsListUri_mock',
  repo: 'reponame_mock',
});