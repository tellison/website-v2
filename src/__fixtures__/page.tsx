import {
    AsciidocPage,
    File,
    MDXPage,
    SingleMDXPage,
    SiteMetaData
} from '../types';

export const createAsciidocData = (): {
    asciidoc: AsciidocPage;
    file: File;
  } => ({
    asciidoc: {
        id: 'asciidoc-1',
        html: `
            <p>Hello World, this is a <pre><code>test</code></pre></p>
            <table><thead><td>Header 1</td><td>Header 2</td></thead><tbody><tr><td>Cell 1</td><td class="icon">Cell 2</td></tr></tbody></table>
            <a href="https://www.eclipse.org">Eclipse</a>
            <span class="icon">[docker]</span>
            <span class="icon">[check]</span>
        `,
        document: {
            title: 'Asciidoc Page title',
            main: 'Asciidoc Page content',
        },
        pageAttributes: {
            authors: 'author1, author2',
        },
        fields: {
            slug: '/asciidoc/asciidoc-page-title',
        }
    },
    file: {
        relativePath: 'test.adoc',
    }
});

export const createMDXData = (): {
    allMdx: MDXPage;
} => ({
    allMdx: {
        edges: [
            {
                node: {
                    frontmatter: {
                        title: 'MDX Page title',
                        author: 'pmc',
                        date: '2021-01-01',
                        describtion: 'MDX Page describtion',
                    },
                    fields: {
                        slug: '/mdx/mdx-page-title',
                        postPath: '/mdx/mdx-page-title',
                    },
                    excerpt: 'MDX Page excerpt',
                }
            }
        ]
    }
});

export const createSingleMDXData = (): {
    mdx: SingleMDXPage;
    site: SiteMetaData;
} => ({
    mdx: {
        id: 'mdx-1',
        excerpt: 'MDX Page excerpt',
        frontmatter: {
            title: 'MDX Page title',
            author: 'pmc',
            date: '2021-01-01',
            description: 'MDX Page describtion',
            tags: ['tag1', 'tag2'],
        }
    },
    site: {
        siteMetadata: {
            title: 'Site title',
            siteUrl: 'https://adoptium.net',
            social: {
                twitter: 'twitter',
            }
        }
    }
});