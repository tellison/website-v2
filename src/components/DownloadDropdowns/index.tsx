import React, { useRef, useState, useCallback, useEffect } from 'react';
import { useLocation } from '@reach/router';
import queryString from 'query-string';
import { Trans } from 'gatsby-plugin-react-i18next';
import VendorSelector from '../VendorSelector'
import { detectOS, UserOS } from '../../util/detectOS';
import { setURLParam } from '../../util/setURLParam';
import { capitalize } from '../../util/capitalize';
import { oses, arches, packageTypes, versions, versionsLTS, defaultVersion, defaultArchitecture, defaultPackageType} from '../../util/defaults';

let defaultOS = 'any'
let defaultArch = 'any'

const DownloadDropdowns = ({updaterAction, marketplace, Table}) => {
    let versionList = versions;
    let selectedVersion = defaultVersion;
    const versionParam = queryString.parse(useLocation().search).version;
    if (versionParam) {
        selectedVersion = Number(versionParam);
    }
    const variantParam = queryString.parse(useLocation().search).variant;
    if (variantParam) {
        // convert openjdk11 to 11
        const parsedVersion = variantParam.toString().replace(/\D/g, '')
        setURLParam('version', parsedVersion)
        selectedVersion = parseInt(parsedVersion);
    }

    if (marketplace) {
        versionList = versionsLTS;
        defaultArch = defaultArchitecture;
        const userOS = detectOS();
        switch (userOS) {
            case UserOS.MAC:
                defaultOS = 'mac'
                if (typeof document !== 'undefined') {
                    let w = document.createElement("canvas").getContext("webgl");
                    let d = w.getExtension('WEBGL_debug_renderer_info');
                    let g = d && w.getParameter(d.UNMASKED_RENDERER_WEBGL) || "";
                    if (g.match(/Apple/) && !g.match(/Apple GPU/)) {
                        defaultArch = 'aarch64'
                    }
                }
                break;
            case UserOS.LINUX:
            case UserOS.UNIX:
                defaultOS = 'linux'
            break;
        default:
            defaultOS = 'windows'
            break;
        }
    }

    const [os, updateOS] = useState(defaultOS);
    const [arch, updateArch] = useState(defaultArch);
    const [packageType, updatePackageType] = useState(defaultPackageType);
    const [version, udateVersion] = useState(selectedVersion);
    
    // Marketplace vendor selector only
    const checkboxRef = useRef({});
    const [checkbox, updateCheckbox] = useState({checkboxRef});

    const [releases, setReleases] = useState(null);

    useEffect(() => {
        (async () => {
          setReleases(await updaterAction(version, os, arch, packageType, checkboxRef));
        })();
    }, [version, os, arch, packageType, checkbox]);

    const setOS = useCallback((os) => {
        updateOS(os);
    }, []);

    const setArch = useCallback((arch) => {
        updateArch(arch);
    }, []);

    const setPackageType = useCallback((packageType) => {
        updatePackageType(packageType);
    }, []);
    
    const setVersion = useCallback((version) => {
        setURLParam('version', version);
        udateVersion(version);
    }, []);

    const setCheckbox= useCallback(() => {
        updateCheckbox({checkboxRef});
    }, []);

    return (
        <>
            {marketplace && <VendorSelector checkboxRef={checkboxRef} setCheckbox={setCheckbox} />}
            <div className="input-group mb-5 row g-2">
                <div className="input-group-prepend flex-colunm col-12 col-md-3">
                    <label className="px-2 fw-bold" htmlFor="os"><Trans>Operating System</Trans></label>
                    <select id="os-filter" aria-label="OS Filter" data-testid="os-filter" onChange={(e) => setOS(e.target.value)} value={os} className="form-select form-select-sm">
                        <option key="any" value="any">Any</option>
                        {oses.map(
                            (os, i): string | JSX.Element => os && (
                                <option key={os.toLowerCase()} value={os.toLowerCase()}>{capitalize(os)}</option>
                            )
                        )}
                    </select>
                </div>
                <div className="input-group-prepend flex-colunm col-12 col-md-3">
                    <label className="px-2 fw-bold" htmlFor="arch"><Trans>Architecture</Trans></label>
                    <select id="arch-filter" aria-label="Architecture Filter" data-testid="arch-filter" onChange={(e) => setArch(e.target.value)} value={arch} className="form-select form-select-sm">
                        <option key="any" value="any">Any</option>
                        {arches.map(
                            (arch, i): string | JSX.Element => arch && (
                                <option key={arch.toLowerCase()} value={arch.toLowerCase()}>{arch}</option>
                            )
                        )}
                    </select>
                </div>
                <div className="input-group-prepend flex-colunm col-12 col-md-3">
                    <label className="px-2 fw-bold" htmlFor="package-type"><Trans>Package Type</Trans></label>
                    <select id="package-type-filter" aria-label="Package Type Filter" data-testid="package-type-filter" onChange={(e) => setPackageType(e.target.value)} value={packageType} className="form-select form-select-sm">
                        <option key="any" value="any">Any</option>
                        {packageTypes.map(
                            (packageType, i): string | JSX.Element => packageType && (
                                <option key={packageType.toLowerCase()} value={packageType.toLowerCase()}>{packageType}</option>
                            )
                        )}
                    </select>
                </div>
                <div className="input-group-prepend flex-colunm col-12 col-md-3">
                    <label className="px-2 fw-bold" htmlFor="version"><Trans>Version</Trans></label>
                    <select id="version-filter" aria-label="Version Filter" data-testid="version-filter" onChange={(e) => setVersion(e.target.value)} value={version} className="form-select form-select-sm">
                        {versionList.map(
                            (version, i): number | JSX.Element => version && (
                                <option key={version} value={version}>{version}</option>
                            )
                        )}
                    </select>
                </div>
            </div>
            <Table results={releases}/>
        </>
    );
};

export default DownloadDropdowns;
