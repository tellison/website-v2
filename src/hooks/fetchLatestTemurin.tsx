import { useEffect, useState } from 'react';
import { VersionMetaData } from ".";

const baseUrl = 'https://api.adoptium.net/v3';

export function fetchLatestForOS(
    isVisible: boolean,
    version: number,
    os: string,
    arch: string
): Binary | null {
    if (!os) {
        return null
    }
    const [binary, setBinary] = useState<Binary | null>(null);
    useEffect(() => {
        if (isVisible) {
        (async () => {
            setBinary(await fetchLatestForOSRequest(version, os, arch));
        })();
        }
    }, [isVisible]);

    return binary;
}

async function fetchLatestForOSRequest(version: number, os: string, arch: string) {
    const url = `${baseUrl}/assets/feature_releases/${version}/ga?os=${os}&architecture=${arch}&image_type=jdk&jvm_impl=hotspot&page_size=1&vendor=eclipse`;
    const response = await fetch(url);
    const json: LatestTemurin = (await response.json())[0];
    let binary_link = json.binaries[0].package.link
    if (json.binaries[0].installer) {
        binary_link = json.binaries[0].installer.link
    }
    return {
        release_name: json.release_name,
        link: binary_link
    };
}

export interface Binary {
  release_name: string;
  link: URL;
}

export interface LatestTemurin {
    download_count: number;
    id: string;
    release_link: URL;
    release_name: string;
    release_type: string;
    source?: {
        link: URL;
        name: string;
        size: number;
    }
    timestamp: Date;
    updated_at: Date;
    vendor: string;
    version_data: VersionMetaData
    binaries: [
        {
            architecture: string;
            download_count: number;
            heap_size: string;
            image_type: string;
            jvm_impl: string;
            os: string;
            package: {
                checksum: string;
                checksum_link: URL;
                download_count: number;
                link: URL;
                metadata_link: URL;
                name: string;
                signature_link: URL;
                size: number;
            }
            installer?: {
                checksum: string;
                checksum_link: URL;
                download_count: number;
                link: URL;
                metadata_link: URL;
                name: string;
                signature_link: URL;
                size: number;
            }
        }
    ]
}