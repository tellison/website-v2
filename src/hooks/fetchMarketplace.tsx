const baseUrl = 'https://marketplace-api.adoptium.net';
import { VersionMetaData } from ".";
    
export async function getAllPkgsForVersion(
    version: number,
    os: string,
    architecture: string,
    package_type: string,
    checkboxRef: any,
): Promise<MarketplaceRelease[] | null> {
    let microsoftSelected = checkboxRef.current.vendorMicrosoft.checked;
    let temurinSelected = checkboxRef.current.vendorAdoptium.checked;
    let redhatSelected = checkboxRef.current.vendorRedHat.checked;
    let huaweiSelected = checkboxRef.current.vendorHuawei.checked;
    let zuluSelected = checkboxRef.current.vendorAzul.checked;
    let ibmSelected = checkboxRef.current.vendorIBM.checked;
    let alibabaSelected = checkboxRef.current.vendorAlibaba.checked;

    let params = '?'
    params += 'feature_version=' + version;

    if (os !== 'any' ) {
        if (os === 'alpine-linux') {
            params += '&os=alpine_linux'
        } else {
            params += ('&os=' + os)
        }
    }

    if (architecture !== 'any' ) {
        params += ('&architecture=' + architecture)
    }

    if (package_type === 'any') {
        params += '&image_type=jdk&image_type=jre'
    } else {
        params += ('&image_type=' + package_type)
    }

    if (temurinSelected) {
        params += '&vendor=adoptium'
    }

    if (redhatSelected) {
        params += '&vendor=redhat'
    }

    if (huaweiSelected) {
        params += '&vendor=huawei'
    }

    if (microsoftSelected) {
        params += ('&vendor=microsoft')
    }

    if (zuluSelected) {
        params += ('&vendor=azul')
    }

    if (ibmSelected) {
        params += ('&vendor=ibm')
    }

    if (alibabaSelected) {
        params += ('&vendor=alibaba')
    }

    const url = new URL(baseUrl + '/v1/assets/latestForVendors' + params);
    const data = await getPkgs(url);
    return data
}

async function getPkgs(url: URL) {
    const response = await fetch(url)
    return response.json();
}

export function getImageForDistribution(distribution: string) {
    switch(distribution) {
        case 'temurin': return '/images/adoptium-logo.png';
        case 'redhat': return '/images/redhat.svg';
        case 'bisheng': return '/images/huawei.svg';
        case 'zulu': return '/images/azul-logo.png';
        case 'semeru': return '/images/ibm-logo.png';
        default: return `/images/${distribution}-logo.png`;
    }
}

export interface MarketplaceRelease {
    release_name: string;
    vendor: string;
    binary: {
        os: string;
        architecture: string;
        image_type: string;
        jvm_impl: string;
        package: {
            name: string;
            link: URL;
            sha265sum: string;
            sha256sum_link: URL;
            signature_link: URL;
        }
        installer?: [{
            name: string;
            link: URL;
            sha265sum: string;
            sha256sum_link: URL;
            signature_link: URL;
        }]
        timestamp: Date;
        scm_ref: string;
        openjdk_scm_ref: string;
        distribution: string;
        aqavit_results_link: URL;
        tck_affidavit_link: URL;
    }
    version: VersionMetaData;
}
