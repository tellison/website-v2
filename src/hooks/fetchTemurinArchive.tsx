import moment from 'moment';
import { VersionMetaData } from '.';
import { fetchExtension } from '../util/fetchExtension';

const baseUrl = 'https://api.adoptium.net/v3';

let releases: TemurinReleases[] = []

export async function getAssetsForVersion(
  version: number,
  releaseType: any,
  numBuilds: number,
  buildDate: Date,
  page: number
): Promise<ReturnedReleases | null> {
  let url = new URL(`${baseUrl}/assets/feature_releases/${version}/${releaseType}?vendor=eclipse`);
  url.searchParams.append('page', page.toString());
  if (numBuilds) {
    url.searchParams.append('page_size', numBuilds.toString());
  }
  if (buildDate) {
    url.searchParams.append('before', moment(buildDate).format('Y-MM-DD'));
  }
  // Expose total page count in header for pagination
  if (releaseType == 'ga') {
    url.searchParams.append('show_page_count', 'true');
  }
  releases = []
  const response = await getPkgs(url)
  const packages = await response.json()
  const pagecount = Number(response.headers.get('pagecount'))
  let pkgsFound: TemurinReleases[] = []
  for (let pkg of packages) {
      pkgsFound.push(pkg);
  }
  return renderReleases(pkgsFound, pagecount);
}

async function getPkgs(url: URL) {
  let response = await fetch(url)
  return response;
}

function renderReleases(pkgs, pagecount) {
  pkgs.forEach((aRelease) => {
    const release: TemurinReleases = {
        release_name: aRelease.release_name,
        release_link: aRelease.release_link,
        timestamp: aRelease.timestamp,
        platforms: {},
    };

    aRelease.binaries.forEach((
      aReleaseAsset: APIResponse) => {
        const platform = `${aReleaseAsset.os}-${aReleaseAsset.architecture}`
  
        // Skip this asset if it's not a binary type we're interested in displaying
        const binary_type = aReleaseAsset.image_type.toUpperCase();
        if (binary_type == 'SOURCES') {
          release.source_url = new URL(aReleaseAsset.package.link);
        }
        if (!['INSTALLER', 'JDK', 'JRE'].includes(binary_type)) {
          return;
        }
  
        if (!release.platforms[platform]) {
          release.platforms[platform] = {
            assets: [],
          };
        }
  
        let binary_constructor: ReleaseAsset = {
          os: aReleaseAsset.os,
          architecture: aReleaseAsset.architecture,
          type: binary_type,
          link: aReleaseAsset.package.link,
          checksum: aReleaseAsset.package.checksum,
          size: Math.floor(aReleaseAsset.package.size / 1000 / 1000),
          extension: fetchExtension(aReleaseAsset.package.name)
        };
  
        if (aReleaseAsset.installer) {
          binary_constructor.installer_link = aReleaseAsset.installer.link;
          binary_constructor.installer_checksum = aReleaseAsset.installer.checksum;
          binary_constructor.installer_size =  Math.floor(aReleaseAsset.installer.size / 1000 / 1000);
          binary_constructor.installer_extension = fetchExtension(aReleaseAsset.installer.name)
        }
  
        // Add the new binary to the release asset
        release.platforms[platform].assets.push(binary_constructor);
    });
    releases.push(release);
  })
  return {releases, pagecount}
}

export interface ReturnedReleases {
  releases: TemurinReleases[];
  pagecount: number;
}

export interface TemurinReleases {
  release_name: string;
  release_link: URL;
  source_url?: URL;
  timestamp: Date;
  platforms: {
    [key: string]: {
      assets: ReleaseAsset[];
    };
  };
}

interface ReleaseAsset {
  os: string;
  architecture: string;
  type: string;
  link: URL;
  checksum: string;
  size: number;
  extension: string;
  installer_link?: URL;
  installer_checksum?: string;
  installer_size?: number;
  installer_extension?: string;
}

interface APIResponse {
  os: string;
  architecture: string;
  image_type: string;
  package: {
    name: string;
    link: URL;
    checksum: string;
    size: number;
  };
  installer?: {
    link: URL;
    name: string;
    checksum: string;
    size: number;
  };
}

export interface MockTemurinFeatureReleaseAPI {
  id: string;
  download_count: number;
  release_name: string;
  release_type: string;
  release_link: URL;
  source?: {
    link: URL;
    name: string;
    size: number;
  };
  binaries: APIResponse[];
  timestamp: Date;
  updated_at: Date;
  vendor: string;
  version_data: VersionMetaData;
}