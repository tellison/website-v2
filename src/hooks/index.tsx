export * from './useOnScreen';
export * from './useAdoptiumContributorsApi';
export * from './fetchMarketplace';
export * from './fetchTemurinReleases';
export * from './fetchTemurinArchive';
export * from './fetchLatestTemurin';
export * from './fetchNews';

export interface VersionMetaData {
    major: number;
    minor: number;
    security: number;
    pre?: string;
    patch: number;
    build: number;
    optional?: string;
    openjdk_version: string;
}