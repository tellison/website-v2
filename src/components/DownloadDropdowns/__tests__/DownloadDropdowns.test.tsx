import React from 'react';
import { render, waitFor, fireEvent } from '@testing-library/react';
import { afterEach, describe, expect, it, vi } from 'vitest'
import { act } from 'react-test-renderer';
import { createRandomTemurinReleases } from '../../../__fixtures__/hooks';
import DownloadDropdowns from '..';

const Table = () => {
  return (
    <div>table</div>
  );
};

// Mock the VendorSelector component as this is tested separately
vi.mock('../../VendorSelector', () => {
  return {
    default: () => <div>vendor-selector</div>,
  };
})

vi.mock('../../../util/defaults', () => {
  return {
    oses: ['mock_os'],
    arches: ['mock_arch'],
    versions: [1],
    defaultVersion: 1,
    versionsLTS: [1],
    defaultPackageType: 'jdk',
    defaultArchitecture: 'mock_arch',
    packageTypes: ['jdk'],
  }
});

afterEach(() => {
  vi.clearAllMocks();
});

const updater = vi.fn().mockImplementation(() => {
  return {
    releases: [
      createRandomTemurinReleases(false),
      createRandomTemurinReleases(true),
    ],
  };
});

vi.mock('query-string', () => ({
  default: {
    parse: () => ({
      version: 8,
      variant: 'openjdk8',
    }),
  }
}));

describe('DownloadDropdowns component', () => {
  it('renders correctly', () => {
    act(() => {
      const { container, getByTestId } = render(
        <DownloadDropdowns
          updaterAction={updater}
          marketplace={false}
          Table={Table}
        />
      );
      waitFor(() => {
        expect(updater).toHaveBeenCalledTimes(1);
      }).then(() => {
        // Simulate a user using dropdowns
        let select = getByTestId("os-filter");
        fireEvent.change(select, { target: { value: "mock_os" } });
        expect(updater).toHaveBeenCalledTimes(2);
        select = getByTestId("arch-filter");
        fireEvent.change(select, { target: { value: "mock_arch" } });
        expect(updater).toHaveBeenCalledTimes(3);
        select = getByTestId("package-type-filter");
        fireEvent.change(select, { target: { value: "any" } });
        expect(updater).toHaveBeenCalledTimes(4);
        select = getByTestId("version-filter");
        fireEvent.change(select, { target: { value: 1 } });
        expect(updater).toHaveBeenCalledTimes(5);
        expect(container).toMatchSnapshot();
      });
    });
  });

  it('renders correctly - marketplace', () => {
    act(() => {
      const { container } = render(
        <DownloadDropdowns
          updaterAction={updater}
          marketplace={true}
          Table={Table}
        />
      );
      waitFor(() => {
        expect(updater).toHaveBeenCalledTimes(1);
      }).then(() => {
        expect(container).toMatchSnapshot();
      });
    });
  });
});