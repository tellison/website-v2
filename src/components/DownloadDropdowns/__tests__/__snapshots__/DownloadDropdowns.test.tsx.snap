// Vitest Snapshot v1

exports[`DownloadDropdowns component > renders correctly - marketplace 1`] = `
<div>
  <div>
    vendor-selector
  </div>
  <div
    class="input-group mb-5 row g-2"
  >
    <div
      class="input-group-prepend flex-colunm col-12 col-md-3"
    >
      <label
        class="px-2 fw-bold"
        for="os"
      >
        Text
      </label>
      <select
        aria-label="OS Filter"
        class="form-select form-select-sm"
        data-testid="os-filter"
        id="os-filter"
      >
        <option
          value="any"
        >
          Any
        </option>
        <option
          value="mock_os"
        >
          Mock_os
        </option>
      </select>
    </div>
    <div
      class="input-group-prepend flex-colunm col-12 col-md-3"
    >
      <label
        class="px-2 fw-bold"
        for="arch"
      >
        Text
      </label>
      <select
        aria-label="Architecture Filter"
        class="form-select form-select-sm"
        data-testid="arch-filter"
        id="arch-filter"
      >
        <option
          value="any"
        >
          Any
        </option>
        <option
          value="mock_arch"
        >
          mock_arch
        </option>
      </select>
    </div>
    <div
      class="input-group-prepend flex-colunm col-12 col-md-3"
    >
      <label
        class="px-2 fw-bold"
        for="package-type"
      >
        Text
      </label>
      <select
        aria-label="Package Type Filter"
        class="form-select form-select-sm"
        data-testid="package-type-filter"
        id="package-type-filter"
      >
        <option
          value="any"
        >
          Any
        </option>
        <option
          value="jdk"
        >
          jdk
        </option>
      </select>
    </div>
    <div
      class="input-group-prepend flex-colunm col-12 col-md-3"
    >
      <label
        class="px-2 fw-bold"
        for="version"
      >
        Text
      </label>
      <select
        aria-label="Version Filter"
        class="form-select form-select-sm"
        data-testid="version-filter"
        id="version-filter"
      >
        <option
          value="1"
        >
          1
        </option>
      </select>
    </div>
  </div>
  <div>
    table
  </div>
</div>
`;

exports[`DownloadDropdowns component > renders correctly 1`] = `
<div>
  <div
    class="input-group mb-5 row g-2"
  >
    <div
      class="input-group-prepend flex-colunm col-12 col-md-3"
    >
      <label
        class="px-2 fw-bold"
        for="os"
      >
        Text
      </label>
      <select
        aria-label="OS Filter"
        class="form-select form-select-sm"
        data-testid="os-filter"
        id="os-filter"
      >
        <option
          value="any"
        >
          Any
        </option>
        <option
          value="mock_os"
        >
          Mock_os
        </option>
      </select>
    </div>
    <div
      class="input-group-prepend flex-colunm col-12 col-md-3"
    >
      <label
        class="px-2 fw-bold"
        for="arch"
      >
        Text
      </label>
      <select
        aria-label="Architecture Filter"
        class="form-select form-select-sm"
        data-testid="arch-filter"
        id="arch-filter"
      >
        <option
          value="any"
        >
          Any
        </option>
        <option
          value="mock_arch"
        >
          mock_arch
        </option>
      </select>
    </div>
    <div
      class="input-group-prepend flex-colunm col-12 col-md-3"
    >
      <label
        class="px-2 fw-bold"
        for="package-type"
      >
        Text
      </label>
      <select
        aria-label="Package Type Filter"
        class="form-select form-select-sm"
        data-testid="package-type-filter"
        id="package-type-filter"
      >
        <option
          value="any"
        >
          Any
        </option>
        <option
          value="jdk"
        >
          jdk
        </option>
      </select>
    </div>
    <div
      class="input-group-prepend flex-colunm col-12 col-md-3"
    >
      <label
        class="px-2 fw-bold"
        for="version"
      >
        Text
      </label>
      <select
        aria-label="Version Filter"
        class="form-select form-select-sm"
        data-testid="version-filter"
        id="version-filter"
      >
        <option
          value="1"
        >
          1
        </option>
      </select>
    </div>
  </div>
  <div>
    table
  </div>
</div>
`;
